[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/drbob-richardson/stat220/blob/main/Assignments/Stat_220_HW9.ipynb)

**Problem 1** A number of measurements were collected for a large variety of cars in order to predict fuel efficiency of the car and to understand what factors significantly affect fuel efficiency. You can read in this data here
cars = pd.read_csv("https://richardson.byu.edu/220/cars.csv")
More information on this data can be found here: https://code.datasciencedojo.com/datasciencedojo/datasets/tree/master/Auto%20MPG
Your task is to build a model that we can use to determine important information about the relationship between fuel efficiency and the other factors.

In [7]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
import statsmodels.formula.api as smf
from statsmodels.stats.outliers_influence import variance_inflation_factor

In [2]:
fuel = pd.read_csv("https://richardson.byu.edu/220/cars.csv")
fuel.head()

mpg  cylinder  displacement  horsepower  weight  acceleration  model_year  \
0  18.0         8         307.0         130    3504          12.0          70   
1  15.0         8         350.0         165    3693          11.5          70   
2  18.0         8         318.0         150    3436          11.0          70   
3  16.0         8         304.0         150    3433          12.0          70   
4  17.0         8         302.0         140    3449          10.5          70   

     origin   car_name  
0  American  chevrolet  
1  American      buick  
2  American   plymouth  
3  American        amc  
4  American       ford

Part a. Build and tune a linear regression model to predict the target variable, mpg or miles per gallon, using all the other variables besides car_name as predictors.  Tune the model so that only significant features remain.

In [9]:
model = smf.ols(
    formula="mpg ~ cylinder + displacement + horsepower + weight + acceleration + model_year + origin",
    data=fuel
).fit()

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                    mpg   R-squared:                       0.824
Model:                            OLS   Adj. R-squared:                  0.821
Method:                 Least Squares   F-statistic:                     224.5
Date:                Thu, 04 Dec 2025   Prob (F-statistic):          1.79e-139
Time:                        22:29:19   Log-Likelihood:                -1020.5
No. Observations:                 392   AIC:                             2059.
Df Residuals:                     383   BIC:                             2095.
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
Intercept            -17.9546      4

In [10]:
model = smf.ols(
    formula="mpg ~ cylinder + displacement + horsepower + weight + model_year + origin",
    data=fuel
).fit()

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                    mpg   R-squared:                       0.824
Model:                            OLS   Adj. R-squared:                  0.821
Method:                 Least Squares   F-statistic:                     256.7
Date:                Thu, 04 Dec 2025   Prob (F-statistic):          1.49e-140
Time:                        22:29:50   Log-Likelihood:                -1020.8
No. Observations:                 392   AIC:                             2058.
Df Residuals:                     384   BIC:                             2089.
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
Intercept            -16.3323      4

In [11]:
model = smf.ols(
    formula="mpg ~ displacement + horsepower + weight + model_year + origin",
    data=fuel
).fit()

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                    mpg   R-squared:                       0.823
Model:                            OLS   Adj. R-squared:                  0.820
Method:                 Least Squares   F-statistic:                     297.9
Date:                Thu, 04 Dec 2025   Prob (F-statistic):          2.80e-141
Time:                        22:30:06   Log-Likelihood:                -1022.0
No. Observations:                 392   AIC:                             2058.
Df Residuals:                     385   BIC:                             2086.
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
Intercept            -17.5036      4

Part b. You will likely find the variable weight significant. What is the confidence interval for the coefficient associated with weight?

In [12]:
model.conf_int().loc["weight"]

0   -0.007693
1   -0.005438
Name: weight, dtype: float64

Part c. You are asked, how does weight affect fuel efficiency. Answer the question while providing some uncertainty to the answer (i.e. interpret the confidence interval).

Cars that weigh more, on average, have a lower fuel efficiency (mpg) than cars that weigh less. This is because in our model the coefficient for weight is negative which shows that as weight increases, predicted average mpg goes down. On average for each increase of 1 in weight this decreases the average predicted mpg by 0.0066. The model predicts with a 95% accuracy that the coefficient will be within -0.00543 and -0.007693.

**Problem 2** Explore adding higher order terms into the model

Part a. Fit and tune a model that includes the interaction between weight and model_year.

In [14]:
model_int = smf.ols(
    formula="mpg ~ cylinder + displacement + horsepower + weight * model_year + acceleration + origin",
    data=fuel
).fit()

print(model_int.summary())

                            OLS Regression Results                            
Dep. Variable:                    mpg   R-squared:                       0.851
Model:                            OLS   Adj. R-squared:                  0.848
Method:                 Least Squares   F-statistic:                     243.3
Date:                Thu, 04 Dec 2025   Prob (F-statistic):          3.27e-152
Time:                        22:51:55   Log-Likelihood:                -987.44
No. Observations:                 392   AIC:                             1995.
Df Residuals:                     382   BIC:                             2035.
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
Intercept           -123.8603     13

In [15]:
model_int = smf.ols(
    formula="mpg ~ displacement + horsepower + weight * model_year + acceleration + origin",
    data=fuel
).fit()

print(model_int.summary())

                            OLS Regression Results                            
Dep. Variable:                    mpg   R-squared:                       0.851
Model:                            OLS   Adj. R-squared:                  0.848
Method:                 Least Squares   F-statistic:                     274.3
Date:                Thu, 04 Dec 2025   Prob (F-statistic):          2.03e-153
Time:                        22:52:12   Log-Likelihood:                -987.50
No. Observations:                 392   AIC:                             1993.
Df Residuals:                     383   BIC:                             2029.
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
Intercept           -124.8020     13

In [16]:
model_int = smf.ols(
    formula="mpg ~ displacement + horsepower + weight * model_year + origin",
    data=fuel
).fit()

print(model_int.summary())

                            OLS Regression Results                            
Dep. Variable:                    mpg   R-squared:                       0.850
Model:                            OLS   Adj. R-squared:                  0.848
Method:                 Least Squares   F-statistic:                     311.8
Date:                Thu, 04 Dec 2025   Prob (F-statistic):          4.13e-154
Time:                        22:52:29   Log-Likelihood:                -988.85
No. Observations:                 392   AIC:                             1994.
Df Residuals:                     384   BIC:                             2025.
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
Intercept           -120.2829     12

Part b. Interpret this model, including the interactions, to a non-technical audience using no numbers.

This model shows that a car’s weight and the year it was made both influence fuel efficiency, and that these two factors work together rather than acting separately. Heavier cars generally get lower gas mileage, but newer cars are built efficiently enough that added weight doesn’t hurt their mpg as much as it does for older cars. Likewise, the advantage of being a newer model is strongest for lighter cars. In short, the impact of weight depends on the model year, and the impact of model year depends on the weight.

Part c. Examine using all higher order terms in the model. Using out of sample metrics, compare a model with no higher terms to a model with all higher order terms (no reducing the full model yet). Remember to keep car name out of the predictor set.

In [18]:
formula_A = "mpg ~ cylinder + displacement + horsepower + weight + acceleration + model_year + origin"
model_A = smf.ols(formula=formula_A, data=fuel).fit()

formula_B = "mpg ~ (cylinder + displacement + horsepower + weight + acceleration + model_year + origin) ** 2 + \
             I(cylinder**2) + I(displacement**2) + I(horsepower**2) + I(weight**2) + I(acceleration**2) + I(model_year**2)"
model_B = smf.ols(formula=formula_B, data=fuel).fit()

In [40]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

train, test = train_test_split(fuel, test_size=0.2, random_state=42)

fit_A = smf.ols(formula_A, data=train).fit()
fit_B = smf.ols(formula_B, data=train).fit()

pred_A = fit_A.predict(test)
pred_B = fit_B.predict(test)

rmse_A = np.sqrt(mean_squared_error(test["mpg"], pred_A))
rmse_B = np.sqrt(mean_squared_error(test["mpg"], pred_B))

print(f"RMSE for model A (no higher-order terms): {rmse_A}")
print(f"RMSE for model B (with higher-order terms): {rmse_B}")
print("AIC for Model A: ", fit_A.aic)
print("AIC for Model B: ", fit_B.aic)

RMSE for model A (no higher-order terms): 3.2561140968477957
RMSE for model B (with higher-order terms): 2.7832516636370124
AIC for Model A:  1651.3383464114147
AIC for Model B:  1529.3919651906956


After comparing rmse for the model without higher-order tems to the model with higher-order terms on the same training and testing data, we conclude that the model with the higher-order terms had a significantly smaller rmse and thus preformed better.

Part d. Reduce the full model to a set of significant predictors. Compare this reduced model to the previous two models (no higher order terms and all higher order terms).

In [29]:
print(model_B.summary())

                            OLS Regression Results                            
Dep. Variable:                    mpg   R-squared:                       0.902
Model:                            OLS   Adj. R-squared:                  0.890
Method:                 Least Squares   F-statistic:                     78.33
Date:                Thu, 04 Dec 2025   Prob (F-statistic):          1.94e-151
Time:                        23:20:50   Log-Likelihood:                -906.47
No. Observations:                 392   AIC:                             1897.
Df Residuals:                     350   BIC:                             2064.
Df Model:                          41                                         
Covariance Type:            nonrobust                                         
                                      coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
Intercept 

In [32]:
formula_reduced = """
mpg ~
    acceleration +
    model_year +
    I(model_year**2) +
    acceleration:origin +
    model_year:origin +
    acceleration:model_year
"""
model_reduced = smf.ols(formula=formula_reduced, data=fuel).fit()

print(model_reduced.summary())

                            OLS Regression Results                            
Dep. Variable:                    mpg   R-squared:                       0.656
Model:                            OLS   Adj. R-squared:                  0.648
Method:                 Least Squares   F-statistic:                     91.15
Date:                Thu, 04 Dec 2025   Prob (F-statistic):           7.51e-84
Time:                        23:22:48   Log-Likelihood:                -1152.2
No. Observations:                 392   AIC:                             2322.
Df Residuals:                     383   BIC:                             2358.
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                                      coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
Intercept 

In [39]:
pred_reduced = model_reduced.predict(test)

rmse_reduced = np.sqrt(mean_squared_error(test["mpg"], pred_reduced))

print(f"RMSE for reduced model: {rmse_reduced}")
print("AIC for Model combined: ", model_reduced.aic)

RMSE for reduced model: 4.303931694613923
AIC for Model combined:  2322.4943436889157


The rmse for the reduced model is significantly higher than the originals. We can conclude that it does not perform as well.

Part e. Explore building a model using the original variables and adding only a few important higher terms you found from the full model. Again compare this model to the others.

In [37]:
formula_combined = """
mpg ~
    displacement +
    acceleration +
    model_year +
    origin +
    I(model_year**2) +
    acceleration:origin +
    model_year:origin +
    acceleration:model_year +
    displacement:model_year
"""

fit_combined = smf.ols(formula_combined, data=train).fit()
pred_combined = fit_combined.predict(test)
rmse_combined = np.sqrt(mean_squared_error(test["mpg"], pred_combined))

print("RMSE for Model combined: ", rmse_combined)
print("AIC for Model combined: ", fit_combined.aic)

RMSE for Model combined:  3.1344186224069888
AIC for Model combined:  1672.0801171261728


We can see that this is a decent model but it doesn't do as well as the original with all of the higher order variables included.

**Problem 3** Use the car data and build a regression tree. Use out of sample metrics to determine the best depth for the tree. Using Feature importances, what variables are considered to be most significant. How does this compare to the linear regression model?

In [41]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error
import numpy as np
import pandas as pd

X_train = pd.get_dummies(train.drop(columns=["mpg", "car_name"]), drop_first=True)
y_train = train["mpg"]

X_test = pd.get_dummies(test.drop(columns=["mpg", "car_name"]), drop_first=True)
y_test = test["mpg"]

X_test = X_test.reindex(columns=X_train.columns, fill_value=0)

depths = range(1, 21)
rmse_values = []

for d in depths:
    tree = DecisionTreeRegressor(max_depth=d, random_state=42)
    tree.fit(X_train, y_train)
    pred = tree.predict(X_test)
    rmse = np.sqrt(mean_squared_error(y_test, pred))
    rmse_values.append(rmse)

best_depth = depths[np.argmin(rmse_values)]
best_depth, rmse_values[np.argmin(rmse_values)]


(6, np.float64(3.0305383555401524))

In [42]:
best_tree = DecisionTreeRegressor(max_depth=best_depth, random_state=42)
best_tree.fit(X_train, y_train)

pred_tree = best_tree.predict(X_test)
rmse_tree = np.sqrt(mean_squared_error(y_test, pred_tree))

print("Best Depth:", best_depth)
print("RMSE for Best Regression Tree:", rmse_tree)

importances = pd.Series(best_tree.feature_importances_, index=X_train.columns)
importances.sort_values(ascending=False).head(10)


Best Depth: 6
RMSE for Best Regression Tree: 3.0305383555401524


displacement       0.653378
horsepower         0.175649
model_year         0.100369
weight             0.053613
acceleration       0.012611
origin_Japanese    0.002263
cylinder           0.002116
origin_European    0.000000
dtype: float64

According to the tree model the displacement is the most important predictor, followed by horsepower, model_year and weight. The rest are not very important.

This is similar to the linear regression model in that weight and model_year seem to be very important. However displacement seems to be much more heavily valued in the tree than the linear regression model. This makes me think that it is important but can't be predicted as well linearly.

**Problem 4**: Consider a new American car with a 6 cylinders, a displacement of 200, a horsepower of 100, a weight of 3600, an acceleration of 14, and a model year of 83.

Part a. For the best regression model in Problem 2, and the best regression tree model in Problem 3, determine the prediction for the new car.

In [48]:
new_car = pd.DataFrame({
    "cylinder": [6],
    "displacement": [200],
    "horsepower": [100],
    "weight": [3600],
    "acceleration": [14],
    "model_year": [83],
    "origin": ["American"],   # reference category
    "car_name": ["newcar"]
})

pred_linear = fit_B.predict(new_car)[0]

X_new = pd.get_dummies(new_car.drop(columns=["mpg","car_name"], errors="ignore"), drop_first=True)
X_new = X_new.reindex(columns=X_train.columns, fill_value=0)

pred_tree = best_tree.predict(X_new)[0]

print(f"Predicted MPG for Linear Regression Model: {pred_linear}")
print(f"Predicted MPG for Regression Tree Model: {pred_tree}")

Predicted MPG for Linear Regression Model: 23.509556313892062
Predicted MPG for Regression Tree Model: 38.0


Part b. Find the confidence interval for the expected value of the prediction in the linear regression model.

In [59]:
pred_mean_ci = fit_B.get_prediction(new_car).summary_frame(alpha=0.05)
pred_mean_ci["mean_ci_lower"], pred_mean_ci["mean_ci_upper"]

(0    19.387543
 Name: mean_ci_lower, dtype: float64,
 0    27.63157
 Name: mean_ci_upper, dtype: float64)

Part c. Find the confidence interval for the predicted value using the linear regression model

In [60]:
pred_pred_ci = fit_B.get_prediction(new_car).summary_frame(alpha=0.05)
pred_pred_ci[["obs_ci_lower", "obs_ci_upper"]]

obs_ci_lower  obs_ci_upper
0     16.911304     30.107809

Part d. Explain why those confidence intervals are different.

The confidence interval for the expected value is narrower because it reflects only the uncertainty in estimating the average mpg for cars with these characteristics. It does not include the natural variation in mpg between individual cars.

Part e. Use bootstrapping to find a confidence interval for the prediction using the regression tree model.

In [61]:
B = 1000

bootstrap_preds = []

for b in range(B):
    idx = np.random.choice(len(X_train), size=len(X_train), replace=True)

    X_boot = X_train.iloc[idx]
    y_boot = y_train.iloc[idx]

    tree_boot = DecisionTreeRegressor(max_depth=best_depth, random_state=42)
    tree_boot.fit(X_boot, y_boot)

    pred_b = tree_boot.predict(X_new)[0]
    bootstrap_preds.append(pred_b)

bootstrap_preds = np.array(bootstrap_preds)

lower = np.percentile(bootstrap_preds, 2.5)
upper = np.percentile(bootstrap_preds, 97.5)

(pred_tree, lower, upper)

(np.float64(38.0), np.float64(17.697499999999998), np.float64(38.0))

the interval is 17.697 - 38

**Problem 5** Airfoils are the part of an airplane wing that allows for stability in flight due to their shape. NASA did a study of the noise level of various shapes and sizes of airfoils.  The data set for their study can be loaded using

airfoil = pd.read_table("https://richardson.byu.edu/220/airfoil.csv")

More information on this data can be found here: https://archive.ics.uci.edu/ml/datasets/airfoil+self-noise. The target variable in the data set is labeled as Pressure. The rest are predictors.

Part a. Build and tune a linear model where you scale and standardize the data.

In [70]:
airfoil = pd.read_csv("https://richardson.byu.edu/220/airfoil.csv")
airfoil.head()

Frequency  Angle  Length  Velocity  Thickness  Pressure
0        800    0.0  0.3048      71.3   0.002663   126.201
1       1000    0.0  0.3048      71.3   0.002663   125.201
2       1250    0.0  0.3048      71.3   0.002663   125.951
3       1600    0.0  0.3048      71.3   0.002663   127.591
4       2000    0.0  0.3048      71.3   0.002663   127.461

In [71]:
model = smf.ols(
    formula="Pressure ~ Frequency + Angle + Length + Velocity + Thickness",
    data=airfoil
).fit()

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:               Pressure   R-squared:                       0.516
Model:                            OLS   Adj. R-squared:                  0.514
Method:                 Least Squares   F-statistic:                     318.8
Date:                Thu, 04 Dec 2025   Prob (F-statistic):          1.15e-232
Time:                        23:53:56   Log-Likelihood:                -4490.1
No. Observations:                1503   AIC:                             8992.
Df Residuals:                    1497   BIC:                             9024.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    132.8338      0.545    243.866      0.0

Part b. The manufacturing company did just build a new airfoil. They want you to use the model to predict the noise level at various velocities and frequencies. The data to predict can be found at

airfoil_new = pd.read_csv("https://richardson.byu.edu/220/airfoil_new.csv")

Find predictions for each of the new rows.

Part c. The manufacturer is exploring using longer airfoil lengths than done before. They want to use lengths of 0.4 meters and they want to use your model to predict the noise level. Are there any issues with that?